### PR TITLE
Modify for getting solutions when setting alpha to 0 with A/B is nullptr

### PR DESCRIPTION
**Pull request by geotseng-amd**
1.  Modify conditions in tensile_host.cpp and rocblaslt_mat_utils.hpp to enable the test case to get solutions when setting alpha to 0 with A/B as nullptr.
2. Add a test case for getting solutions when setting alpha to 0 with A/B as nullptr
3. Ticket: https://ontrack-internal.amd.com/browse/SWDEV-474250

### CODE REVIEW

**Review by TonyYHsieh — APPROVED** [review]
LGTM

File: clients/include/testing_auxiliary.hpp [hunk 1]
```diff
@@ -392,6 +392,93 @@ void testing_aux_get_sol_with_null_biasaddr(const Arguments& arg)
...
+                                                          matmul,
+                                                          matA,
+                                                          matB,
+                                                          matC,
+                                                          matD,
+                                                          pref,
+                                                          request_solutions,
+                                                          heuristicResult,
+                                                          &returnedAlgoCount));
+
```
Comment (jichangjichang) [inline clients/include/testing_auxiliary.hpp:462]: it is better to try run the solution as well. (hipblasltMatmul())

File: clients/gtest/auxiliary_gtest.yaml [hunk 2]
```diff
@@ -84,6 +84,11 @@ Tests:
   function:
     - aux_get_sol_with_null_biasaddr: *hpa_half_precision
 
+- name: aux_get_sol_with_zero_alpha_null_a_b
+  category: pre_checkin
+  function:
```
Comment (jichangjichang) [inline clients/gtest/auxiliary_gtest.yaml:89]: The original request is for cpp path.
It is better to add test for cpp path as well.

File: clients/include/testing_auxiliary.hpp (hunk 1 above)
Comment (geotseng-amd) [reply clients/include/testing_auxiliary.hpp:462]: Done

File: clients/include/testing_auxiliary.hpp [hunk 3]
```diff
@@ -392,6 +392,111 @@ void testing_aux_get_sol_with_null_biasaddr(const Arguments& arg)
...
+        hipblasLtMatmulDescCreate(&matmul, arg.compute_type, arg.scale_type));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmulDescSetAttribute(
+        matmul, HIPBLASLT_MATMUL_DESC_TRANSA, &trans_a, sizeof(int32_t)));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmulDescSetAttribute(
+        matmul, HIPBLASLT_MATMUL_DESC_TRANSB, &trans_b, sizeof(int32_t)));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmulDescSetAttribute(
+        matmul, HIPBLASLT_MATMUL_DESC_EPILOGUE, &epilogue,sizeof(epilogue)));
+
+    // Validation for solution running.
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmul(handle,
```
Comment (jichangjichang) [inline clients/include/testing_auxiliary.hpp]: hipblasLTMatmul() should use the solution returned from hipblasLtMatmulAlgoGetHeuristic()

File: clients/include/testing_auxiliary.hpp [hunk 4]
```diff
@@ -392,6 +393,201 @@ void testing_aux_get_sol_with_null_biasaddr(const Arguments& arg)
...
+                                          d_c,
+                                          matC,
+                                          d_d,
+                                          matD,
+                                          &heuristicResult[0].algo,
+                                          d_workspace,
+                                          max_workspace_size,
+                                          stream));
+
+    CHECK_HIP_ERROR(hipFree(a));
```
Comment (jichangjichang) [inline clients/include/testing_auxiliary.hpp]: Don't free nullptr

File: clients/include/testing_auxiliary.hpp [hunk 5]
```diff
@@ -392,6 +393,201 @@ void testing_aux_get_sol_with_null_biasaddr(const Arguments& arg)
...
+    int64_t            batch_count = 1;
+    // Setting alpha = 0.
+    float              alpha = 0;
+    float              beta = arg.beta;
+    // Setting d_a, d_b, a, b as nullptr.
+    void*              d_a = NULL;
+    void*              d_b = NULL;
+    void*              d_c;
+    void*              d_d;
+    void*              a = NULL;
```
Comment (jichangjichang) [inline clients/include/testing_auxiliary.hpp]: host a ptr and  host b ptr are  not used.

File: library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp [hunk 6]
```diff
@@ -184,7 +184,8 @@ inline rocblaslt_status validateMatmulArgs(int64_t     m,
         return rocblaslt_status_invalid_pointer;
 
     // pointers must be valid
-    if(n && ((k && (!a || !b || !alpha)) || !c || !d))
+    // Update for the valid case: (alpha=0 && (A=NULL || B=NULL))
+    if(n && ((k && (!alpha || ((*((float*)alpha)) && (!a || !b)))) || !c || !d))
```
Comment (TonyYHsieh) [inline library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp:188]: can we split those case to multiple line??  it is too complicate to read.

File: library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp [hunk 7]
```diff
@@ -1063,8 +1063,9 @@ rocblaslt_status rocblaslt_matmul(rocblaslt_handle             handle,
     }
 
     // Check if pointer is valid
-    if(alpha == nullptr || beta == nullptr || A == nullptr || B == nullptr || C == nullptr
-       || D == nullptr)
+    // Update for the valid case: (alpha=0 && (A=NULL || B=NULL))
+    if(alpha == nullptr || beta == nullptr || ((*((float*)alpha)) && (A == nullptr
+        || B == nullptr)) || C == nullptr || D == nullptr)
```
Comment (TonyYHsieh) [inline library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp]: can we split those case to multiple line??  it is too complicate to read

File: library/src/amd_detail/rocblaslt/src/tensile_host.cpp [hunk 8]
```diff
@@ -1482,8 +1482,9 @@ rocblaslt_status gemmCreate(RocblasltContractionProblem const& problem,
     try
     {
         // Check if pointer is valid
-        if(problem.alpha == nullptr || problem.beta == nullptr || problem.A == nullptr
-           || problem.B == nullptr || problem.C == nullptr || problem.D == nullptr)
+        // Update for the valid case: (alpha=0 && (A=NULL || B=NULL))
+        if(problem.alpha == nullptr || problem.beta == nullptr || ((*((float*)problem.alpha)) && (problem.A == nullptr
+           || problem.B == nullptr)) || problem.C == nullptr || problem.D == nullptr)
```
Comment (TonyYHsieh) [inline library/src/amd_detail/rocblaslt/src/tensile_host.cpp]: can we split those case to multiple line??  it is too complicate to read

File: clients/include/testing_auxiliary.hpp [hunk 9]
```diff
@@ -392,6 +393,201 @@ void testing_aux_get_sol_with_null_biasaddr(const Arguments& arg)
...
+    CHECK_HIP_ERROR(hipStreamCreate(&stream));
+    CHECK_HIPBLASLT_ERROR(hipblasLtCreate(&handle));
+    CHECK_HIP_ERROR(hipMalloc(&d_c, m * n * batch_count * sizeof(OutType)));
+    CHECK_HIP_ERROR(hipMalloc(&d_d, m * n * batch_count * sizeof(OutType)));
+    CHECK_HIP_ERROR(hipHostMalloc(&c, m * n * batch_count * sizeof(OutType)));
+    CHECK_HIP_ERROR(hipHostMalloc(&d, m * n * batch_count * sizeof(OutType)));
+    // Allocate space for max_workspace_size.
+    CHECK_HIP_ERROR(hipMalloc(&d_workspace, max_workspace_size));
+    
+    CHECK_HIP_ERROR(hipMemcpyAsync(
```
Comment (jichangjichang) [inline clients/include/testing_auxiliary.hpp]: copying un-initialized data is useless?

File: clients/include/testing_auxiliary.hpp [hunk 10]
```diff
@@ -392,6 +393,201 @@ void testing_aux_get_sol_with_null_biasaddr(const Arguments& arg)
...
+
+    CHECK_HIP_ERROR(hipStreamCreate(&stream));
+    CHECK_HIPBLASLT_ERROR(hipblasLtCreate(&handle));
+    CHECK_HIP_ERROR(hipMalloc(&d_c, m * n * batch_count * sizeof(OutType)));
+    CHECK_HIP_ERROR(hipMalloc(&d_d, m * n * batch_count * sizeof(OutType)));
+    CHECK_HIP_ERROR(hipHostMalloc(&c, m * n * batch_count * sizeof(OutType)));
+    CHECK_HIP_ERROR(hipHostMalloc(&d, m * n * batch_count * sizeof(OutType)));
+    // Allocate space for max_workspace_size.
+    CHECK_HIP_ERROR(hipMalloc(&d_workspace, max_workspace_size));
+
```
Comment (TonyYHsieh) [inline clients/include/testing_auxiliary.hpp]: clean space when you commit

File: clients/include/testing_auxiliary.hpp [hunk 11]
```diff
@@ -392,6 +393,201 @@ void testing_aux_get_sol_with_null_biasaddr(const Arguments& arg)
...
+    float              beta = arg.beta;
+    // Setting d_a, d_b, a, b as nullptr.
+    void*              d_a = NULL;
+    void*              d_b = NULL;
+    void*              d_c;
+    void*              d_d;
+    void*              a = NULL;
+    void*              b = NULL;
+    void*              c;
+    void*              d;
```
Comment (TonyYHsieh) [inline clients/include/testing_auxiliary.hpp]: remove all host memory part.

File: clients/include/testing_auxiliary.hpp [hunk 12]
```diff
@@ -392,6 +393,201 @@ void testing_aux_get_sol_with_null_biasaddr(const Arguments& arg)
...
```
Comment (TonyYHsieh) [inline clients/include/testing_auxiliary.hpp]: remove all host memory part.

File: clients/include/testing_auxiliary.hpp [hunk 13]
```diff
@@ -392,6 +393,201 @@ void testing_aux_get_sol_with_null_biasaddr(const Arguments& arg)
...
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmulDescSetAttribute(
+        matmul, HIPBLASLT_MATMUL_DESC_TRANSB, &trans_b, sizeof(int32_t)));
+
+    hipblasLtEpilogue_t epilogue = HIPBLASLT_EPILOGUE_DEFAULT;
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmulDescSetAttribute(
+        matmul, HIPBLASLT_MATMUL_DESC_EPILOGUE, &epilogue,sizeof(epilogue)));
+
+    // Set User Preference attributes
+    hipblasLtMatmulPreference_t pref;
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmulPreferenceCreate(&pref));
```
Comment (TonyYHsieh) [inline clients/include/testing_auxiliary.hpp:447]: should we care workspace in pref?? you do it in ext but not in here.

**Review by TonyYHsieh — CHANGES_REQUESTED** [review]

File: clients/include/testing_auxiliary.hpp (hunk 13 above)
Comment (jichangjichang) [reply clients/include/testing_auxiliary.hpp:447]: I agreed. This test focus on null a/b. I don't think we need to set and allocate workspace for this test.

**Review by TonyYHsieh — APPROVED** [review]
LGTM

**Review by jichangjichang — APPROVED** [review]
lgtm

**Comment by jichangjichang** [comment]
please use "squash and merge" and don't use "Rebase and merge"

**Comment by geotseng-amd** [comment]
Local test results for PR#956 are all passed.

[0;32m[----------] [mGlobal test environment tear-down
[0;32m[==========] [m35004 tests from 13 test suites ran. (1879589 ms total)
[0;32m[  PASSED  ] [m35004 tests.
hipBLASLt version: 800

**Pull request merged**